File: python/tests/io_expander/pcf8574_test_unit.py
import sys

from periph.connection.i2c_mock import I2CConnectionMock
from periph.chips.io_expander.pcf8574 import Pcf8574Full

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# PCF8574 has no sub-registers: every transaction is a single plain byte
# read()/write() (no register pointer), so I2CConnectionMock's register map
# is never consulted — reads must be preloaded via queue_read() in the
# exact order the driver will issue them. Pcf8574Full's constructor issues
# one extra read (to seed `prev` for interrupt comparison) right after
# Pcf8574Minimal's init write, so it must be queued too.
connection = I2CConnectionMock()
connection.queue_read([0xFF])  # Full.__init__ seeds `prev` via read_port_raw()
chip = Pcf8574Full(connection)
check_true('init', True)

# Construction writes 0xFF (all pins to quasi-bidirectional input mode).
check_true('init_writes_0xff', connection.writes[0] == bytes([0xFF]))
check_true('init_shadow', chip._shadow == 0xFF)

# read_port(): plain single-byte read.
connection.queue_read([0x5A])
check_true('read_port', chip.read_port() == 0x5A)

# write_port(): plain single-byte write; updates shadow.
chip.write_port(0, 0x3C)
check_true('write_port', connection.writes[-1] == bytes([0x3C]))
check_true('write_port_shadow', chip._shadow == 0x3C)

# pin().value() reads the live bus level (not the shadow).
pin3 = chip.pin(3)
connection.queue_read([0x08])  # bit 3 high
check_true('pin_read', pin3.value() == 1)

# Pin set high/low preserves other shadow bits (read-modify-write).
chip.write_port(0, 0xFF)
pin3.off()
check_true('pin3_off', connection.writes[-1] == bytes([0xFF & ~0x08]))
pin5 = chip.pin(5)
pin5.off()
check_true('pin5_off_preserves_pin3', connection.writes[-1] == bytes([0xFF & ~0x08 & ~0x20]))
pin3.on()
check_true('pin3_on_preserves_pin5', connection.writes[-1] == bytes([0xFF & ~0x20]))

# Toggle.
pin3.toggle()
check_true('pin3_toggle_off', connection.writes[-1] == bytes([0xFF & ~0x20 & ~0x08]))
pin3.toggle()
check_true('pin3_toggle_on', connection.writes[-1] == bytes([0xFF & ~0x20]))

# init(IN)/init(OUT) sets the pin high (input/quasi-bidirectional) or low.
pin0 = chip.pin(0)
pin0.init(Pcf8574Full.OUT)
check_true('pin_init_out_drives_low', (connection.writes[-1][0] & 0x01) == 0)
pin0.init(Pcf8574Full.IN)
check_true('pin_init_in_releases_high', (connection.writes[-1][0] & 0x01) == 1)

# Full: poll_interrupt() compares to the previous read and returns the
# changed-pin bitmask, also updating the stored previous value.
chip._prev = 0xFF
connection.queue_read([0xF7])  # bit 3 now low
check_true('poll_interrupt_detects_change', chip.poll_interrupt() == 0x08)
connection.queue_read([0xF7])  # no further change
check_true('poll_interrupt_no_change', chip.poll_interrupt() == 0x00)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
sys.exit(0 if failed == 0 else 1)

File: nodejs/tests/io_expander/pcf8574_test_unit.js
'use strict';
const { I2CConnectionMock } = require('../../packages/periph/src/connection/i2c_mock');
const { Pcf8574Full } = require('../../packages/periph/src/chips/io_expander/pcf8574');

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log('FAIL', label); failed++; }
}

function flushMicrotasks() {
    return new Promise((resolve) => setImmediate(resolve));
}

async function main() {
    // PCF8574 has no sub-registers: every transaction is a single plain
    // byte read()/write() (no register pointer), so I2CConnectionMock's
    // register map is never consulted — reads must be preloaded via
    // queueRead() in the exact order the driver will issue them.
    const connection = new I2CConnectionMock();
    const chip = new Pcf8574Full(connection);
    await flushMicrotasks(); // let the fire-and-forget constructor write (and prev-seed read) finish
    checkTrue('init', true);

    // Construction writes 0xFF (all pins to quasi-bidirectional input mode).
    checkTrue('init_writes_0xff', connection.writes[0].length === 1 && connection.writes[0][0] === 0xFF);
    checkTrue('init_shadow', chip._shadow === 0xFF);

    // readPort(): plain single-byte read.
    connection.queueRead([0x5A]);
    checkTrue('read_port', (await chip.readPort()) === 0x5A);

    // writePort(): plain single-byte write; updates shadow.
    await chip.writePort(0, 0x3C);
    checkTrue('write_port', connection.writes[connection.writes.length - 1][0] === 0x3C);
    checkTrue('write_port_shadow', chip._shadow === 0x3C);

    // pin().read() reads the live bus level (not the shadow).
    const pin3 = chip.pin(3);
    connection.queueRead([0x08]); // bit 3 high
    checkTrue('pin_read', (await pin3.read()) === 1);

    // Pin set high/low preserves other shadow bits (read-modify-write).
    await chip.writePort(0, 0xFF);
    await pin3.write(0);
    checkTrue('pin3_off', connection.writes[connection.writes.length - 1][0] === (0xFF & ~0x08 & 0xFF));
    const pin5 = chip.pin(5);
    await pin5.write(0);
    checkTrue('pin5_off_preserves_pin3', connection.writes[connection.writes.length - 1][0] === (0xFF & ~0x08 & ~0x20 & 0xFF));
    await pin3.write(1);
    checkTrue('pin3_on_preserves_pin5', connection.writes[connection.writes.length - 1][0] === (0xFF & ~0x20 & 0xFF));

    // Toggle is manual (no toggle() method on Node's Pin) — set/read directly.
    // setDirection('in') releases high; setDirection('out') would drive low
    // without a value, but the driver's write() is what actually drives level.
    await pin3.setDirection('out');
    checkTrue('set_direction_out', pin3.direction === 'out');
    await pin3.setDirection('in');
    checkTrue('set_direction_in_releases_high', connection.writes[connection.writes.length - 1][0] & 0x08);

    // Full: pollInterrupt() compares to the previous read and returns the
    // changed-pin bitmask, also updating the stored previous value.
    connection.queueRead([0xFF]);
    await chip.pollInterrupt(); // resync _prev to a known value (0xFF)
    connection.queueRead([0xF7]); // bit 3 now low
    checkTrue('poll_interrupt_detects_change', (await chip.pollInterrupt()) === 0x08);
    connection.queueRead([0xF7]); // no further change
    checkTrue('poll_interrupt_no_change', (await chip.pollInterrupt()) === 0x00);

    console.log(`===DONE: ${passed} passed, ${failed} failed===`);
    process.exit(failed === 0 ? 0 : 1);
}

main();
